### PR TITLE
Chapter? Bayesian Optimization

### PR DESCRIPTION
**Pull request by sumny**

### CODE REVIEW

**Comment by sumny** [comment]
- [ ] move chapter after the beyond chapter

**Comment by sumny** [comment]
@RaphaelS1 as I started pretty late with drafting this chapter, Bernd and I had review meeting prior to you writing a review (also due to Bernds holiday).
Nevertheless, If you find the time, I would love to get a quick review from you.
Contentwise, @mb706 and @be-marc will also do a pass.

**Comment by RaphaelS1** [comment]
Sure will review in Friday!

**Comment by sumny** [comment]
> Sure will review in Friday!

@RaphaelS1 ping :)

**Comment by RaphaelS1** [comment]
@sumny  (CC @berndbischl ) I'm putting reviewing this on hold for the moment as I have some things to discuss with other editors. Please don't merge it but I haven't forgotten about it

File: book/_quarto.yml
Comment (RaphaelS1) [inline book/_quarto.yml:39]: Obviously will need to think more carefully about where this goes in the long run

File: book/_setup.qmd
Comment (RaphaelS1) [inline book/_setup.qmd:7]: This shouldn't be in the general setup file. Please move to start of this chapter instead

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:22]: Update once chapter position fixed

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:30]: Some of this is already covered in C4, as you note below, so would instead linking back to C4 in the first line and then generalising this to black-box more generally

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:38]:
```suggestion
In general, most black box optimizers work iteratively, i.e., they propose new points for evaluation by making use of the information collected during the evaluation of previous points.
```

Comment (RaphaelS1) [reply book/bayesian_optimization.qmd:38]: I think this might have been covered in C4

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:44]: I think it needs to be clearer why we are including this chapter before going into some complex terminology. Personally this is an area I'm very interested in learning more about, but this intro makes it seem complicated and doesn't tell me in simple terms *why* I should care

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:46]: As above, this is a bit too complicated for the 'why'

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:61]: Would suggest rephrasing to make simpler as: i) in C4 we saw classes Tuner...; ii) these are special cases of Optim... from bbotk.

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:64]: Reference figure below here

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:65]: Reference fig below, point to local minima and briefly explain why this means algs might get stuck

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:67]: Broken link to sec-defining...

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:109]: What does this function do?

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:131]: link to GH

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:135]: what's a surrogate?

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:139]: Sorry but this is too complicated to follow. Either break it up with examples, use some maths, or clearly define terms upfront before discussing how they work

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:147]: I don't fully know what the 'note' is

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:148]: New line

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:154]:
```suggestion
The initial design refers to the first set of points the black box function was evaluated on before BO algorithms start their loop.
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:233]: I don't fully understand this description

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:247]: Can you give a concrete example in code, pseudo-code, or as a figure?

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:275]:
```suggestion
* A Gaussian process (`r ref("mlr_learners_regr.km")`) for low dimensional numeric search spaces
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:276]:
```suggestion
* A random forest for higher dimensional mixed (and / or hierarchical) search spaces
```
We implement several so shouldn't direct users to just one

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [reply book/bayesian_optimization.qmd:275]: we're removing the "text" argument from `ref`

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:278]:
```suggestion
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:284]:
```suggestion
surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS",
  control = list(trace = FALSE)))
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:291]:
```suggestion
surrogate = srlrn(lrn("regr.km", covtype = "matern5_2", optim.method = "BFGS",
  control = list(trace = FALSE)), archive = instance$archive)
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:267]: Whole section a bit hard to follow, don't have a good suggestion right now for how best to improve but basically there's just a lot of new and technical content

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:338]: This is great

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:436]: I think the problem with this chapter is that it takes quite a long time to get to, IMO, the most important part, which is this section. Here we finally see why we should care about everything above. I think it could benefit from having an example like this very early on to hook users

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:609]: On first read unclear what this section adds that hasn't been covered in 12.4 and C4

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:729]:
```suggestion
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:713]: Hard to follow without an example

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:761]:
```suggestion
## Practical considerations in Bayesian Optimization
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:814]: Unsure how much of this detail is really needed and useful here, it is very technical. On the other hand some readers might really like that

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:820]:
```suggestion
Bayesian optimization using `r mlr3mbo` relies on a `r ref("Surrogate")`, `r ref("AcqFunction")` and `r ref("AcqOptimizer")` as well as a general `r ref("loop_function")` that build the actual optimizer or tuner constructed in the form of an `r ref("OptimizerMbo")` or `r ref("TunerMbo")`.
```

File: book/bayesian_optimization.qmd
Comment (RaphaelS1) [inline book/bayesian_optimization.qmd:821]: link to vignette

**Review by RaphaelS1 — COMMENTED** [review]
Overall thoughts are that it's a very interesting chapter and one that I look forward to learning from - but currently it's quite complicated and therefore hard to follow, but well-written overall

**Comment by sumny** [comment]
@larskotthoff you can now start your review!

**Comment by larskotthoff** [comment]
- [ ] `set.seed()` doesn't need to be visible in book.
- [ ] 12.1 Ref #sec-defining-search-spaces is broken.
- [ ] 12.1 Explanation of the objective function: what are the fundamental differences between ObjectiveRFun etc? Can you do something with one that you can't do with the other? The text at the moment makes it sound like the only difference is efficiency.
- [ ] After Figure 12.1, minimum value is given -- would be great to show the code to get it from `xydt`.
- [ ] End of 12.1, list of optimizers -- might be worth pointing out that MBO is BO as described in this chapter.
- [ ] 12.2.1 "One can simply evaluate points on the OptimInstance that is to be optimized prior to using an OptimizerMbo." -- don't understand this sentence. It sounds like I need to run an optimizer before running the optimizer, but I guess you mean just starting without an initial design? I.e. the initial surrogate model is just the prior and not based on any observed data?
- [ ] 12.2.1 "must not guarantee" -> "does not guarantee"
- [ ] 12.2.1 would be great to at the end explicitly say how this initial design is given to the optimizer -- the code example implicitly gives one, but what if I want to use `bb_optimize()`?
- [ ] 12.2.3 first code example explanation says that we're optimizing the marginal likelihood -- of what, and where can I see that in the code?
- [ ] 12.2.3 the "using surrogate learner interactively" code example looks exactly like the one above, except for the additional argument -- so was the previous code example wrong? How do I otherwise tell the instance about the surrogate learner? Do I pass the constructed `surrogate` to an `OptimizerMbo`? Could be forward ref to 12.3.
- [ ] Figure 12.3 would be great to show the code for this.
- [ ] 12.2.4 not sure how useful it is to show how to compute EI analytically.
- [ ] Figure 12.4 also would be great to show the code for generating it.
- [ ] 12.2.5 a short note to say why we need an optimizer (i.e. evaluating all possible values is too expensive, even on the surrogate model) would be useful.
- [ ] 12.3 would be good to include an initial design in the example at the beginning.
- [ ] 12.3 what is the `loop_function` here, or is it not necessary? It's prominently mentioned when talking about the individual pieces, so it would be good to refer back to it when putting everything together.

- [ ] In general it would be good to have an explicit example of a `loop_function` somewhere (probably in multiple places). It feels vague at the moment because the chapter never says anywhere "and here is the `loop_function` in this case".

**Comment by sumny** [comment]
@larskotthoff thanks for the comments! I addressed all of them. Can you recheck if you believe that the issues have been resolved? If so, we can probably merge it and ping Raphael.
Below I replied to the issues where I felt need some more elaboration.
All other ones are fixed without any comment from my side!

> 12.1 Explanation of the objective function: what are the fundamental differences between ObjectiveRFun etc? Can you do something with one that you can't do with the other? The text at the moment makes it sound like the only difference is efficiency.

`ObjectiveRFun` wraps a function operating on a single configuration, `ObjectiveRFunMany` wraps a function operating on a list of configurations and `ObjectiveRFunDt` wraps a function directly operating on a `data.table`.
All objectives essentially "do" the same thing (they all have `$eval()`, `$eval_many()` and `$eval_dt()` methods) but differ in how efficiently they do things w.r.t data type conversions and the potential of parallel execution.
I rephrased parts of that description!

 > 12.2.1 "One can simply evaluate points on the OptimInstance that is to be optimized prior to using an OptimizerMbo." -- don't understand this sentence. It sounds like I need to run an optimizer before running the optimizer, but I guess you mean just starting without an initial design? I.e. the initial surrogate model is just the prior and not based on any observed data?

The idea is that before `OptimizerMbo` does anything on the instance you can simply evaluate your own custom `design` via `instance$eval_batch(design)` and the `loop_function` passed to the `OptimizerMbo` will then not create an initial design itself because there are already evaluations in the archive. I rephrased this and I believe this is easier to understand now when also looking at the `loop_function` I included.

>  12.2.1 would be great to at the end explicitly say how this initial design is given to the optimizer -- the code example implicitly gives one, but what if I want to use bb_optimize()?

I have now rephrased how the whole inital design stuff works:
If a specific initial design is desired, it needs to be evaluated on the `OptimInstance` before running an `OptimizerMbo`.
The loop function within the `OptimizerMbo` will then recognize the evaluations in the instance's archive and consider it as the initial design, omitting the need to construct and evaluate a new initial design.
Note that the same mechanism also allows for specifying a custom initial design (i.e., points designed by the user) in the form of a `data.table`.
This holds regardless of whether you use an `OptimizerMbo` or `bb_optimizer()` also note that 

>  12.2.3 first code example explanation says that we're optimizing the marginal likelihood -- of what, and where can I see that in the code?

This was referring to how the GP finds its optimal kernel parameters during training. I rephrased this because it is not that important here.

> 12.2.3 the "using surrogate learner interactively" code example looks exactly like the one above, except for the additional argument -- so was the previous code example wrong? How do I otherwise tell the instance about the surrogate learner? Do I pass the constructed surrogate to an OptimizerMbo? Could be forward ref to 12.3.

Indeed the only difference to "interactive" usage is that you must then supply the archive of the instance during construction of the surrogate. When used within an OptimizerMbo, this initialization step is completed within the loop function which becomes more clear in the provided loop function example.

>  12.3 what is the loop_function here, or is it not necessary? It's prominently mentioned when talking about the individual pieces, so it would be good to refer back to it when putting everything together.

I believe this is now explained in detail in 12.2.2. I also state this in this section now.

>  In general it would be good to have an explicit example of a loop_function somewhere (probably in multiple places). It feels vague at the moment because the chapter never says anywhere "and here is the loop_function in this case".

I included this now in 12.2.2.

**Review by larskotthoff — APPROVED** [review]

**Comment by larskotthoff** [comment]
Thanks, looks good!

**Comment by sumny** [comment]
@RaphaelS1 I guess we can then merge this now and hand it over to you!

**Pull request merged**
